### PR TITLE
Add SecurityCodeService.

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import deepEqual from 'deep-equal';
+import { InvalidAddress } from '../utils/errors';
+
+export class SecurityCodeService {
+  maxCodeNumber: number;
+  securityCodeLength: number;
+  securityKeyboardSize: number;
+  mask: number;
+  highNumberMask: number;
+  lowNumberMask: number;
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]:
```
    maxCodeNumber: number = 1024;
    securityCodeLength: number = 6;
    securityKeyboardSize: number = 30;
    mask: number = 0x03FF;
    highNumberMask: number = 0x03;
    lowNumberMask: number = 0xFF;
```

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    this.maxCodeNumber = 1024;
+    this.securityCodeLength = 6;
+    this.securityKeyboardSize = 30;
+    this.mask = 0x03FF;
+    this.highNumberMask = 0x03;
+    this.lowNumberMask = 0xFF;
+  }
+
+  encode(address: string): number[] {
+    if (!address.match(/0x[0-9A-Fa-f]{40}/)) {
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]: extract function `isProperAddress`

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    const addressBytes = utils.arrayify(address);
+    const code = [];
+    for (let i = 0; i !== addressBytes.length; i += 2) {
+      code.push(this.to10bitNumber(addressBytes[i], addressBytes[i + 1]));
+    }
+
+    return code
+      .slice(0, this.securityCodeLength)
+      .map((codeElement) => codeElement & 0x03FF);
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]: `.map((codeElement) => codeElement & 0x03FF)` 
-> 
`.map((codeElement) => codeElement & this.mask)`

File: universal-login-sdk/lib/services/SecurityCodeService.ts (hunk 2 above)
Comment (JustynaBroniszewska) [reply universal-login-sdk/lib/services/SecurityCodeService.ts]: use `ensure`

File: universal-login-sdk/test/services/SecurityCodeService.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      expect(securityCodeService.to10bitNumber(high, low)).to.eq(longNumber);
+    });
+
+    it('trunked 8bit -> 10bit conversion', () => {
+      expect(securityCodeService.to10bitNumber(1024 + high, 2048 + low)).to.eq(longNumber);
+    });
+  });
+
+  it('argument should be valid address', () => {
+    const invalidAddress1 = mockedAddress.slice(0, 16);
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/test/services/SecurityCodeService.ts]: This should be tests of `isProperAddress`

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import { utils } from 'ethers';
+import deepEqual from 'deep-equal';
+import { InvalidAddress } from '../utils/errors';
+
+export class SecurityCodeService {
+  maxCodeNumber: number;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]: `alfabetSize`

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  securityCodeLength: number;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]: codeLength

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  securityKeyboardSize: number;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]: ?

File: universal-login-sdk/lib/services/SecurityCodeService.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  encode(address: string): number[] {
+    if (!address.match(/0x[0-9A-Fa-f]{40}/)) {
+      throw new InvalidAddress(address);
+    }
+    const addressBytes = utils.arrayify(address);
+    const code = [];
+    for (let i = 0; i !== addressBytes.length; i += 2) {
+      code.push(this.to10bitNumber(addressBytes[i], addressBytes[i + 1]));
+    }
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/services/SecurityCodeService.ts]: `generateCode(address)`

Comment (marekkirejczyk) [reply universal-login-sdk/lib/services/SecurityCodeService.ts]:
```ts

```

File: universal-login-commons/test/core/utils/hexStrings.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  });
+
+  describe('reverse hex string', () => {
+    const hexString = '0xFFFFFFe7d45c34110B34Ed269AD86248884E78C7';
+
+    it('empty hex 0x', () => {
+      expect(reverseHexString('0x')).to.eq('0x');
+    });
+
+    it('0xab == rev(0xba', () => {
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/hexStrings.ts]: Missing )

File: universal-login-commons/test/core/utils/hexStrings.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      expect(reverseHexString('0xab')).to.eq('0xba');
+    });
+
+    it('rev(rev(hex)) == hex', () => {
+      expect(reverseHexString(reverseHexString(hexString))).to.eq(hexString);
+    });
+
+    it('should fail', () => {
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/hexStrings.ts]: not sure what this one is testing

File: universal-login-sdk/test/services/SecurityCodeService.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      const encoding = securityCodeService.generateCode(mockedAddress);
+      expect(encoding).to.deep.eq([ 619, 192, 20, 934, 264, 392 ]);
+    });
+
+    it('returns 6 numbers', () => {
+      const encoding = securityCodeService.generateCode('0x0aFFFFe7d45c34110B34Ed269AD86248884E78C7');
+      expect(encoding).to.deep.eq([350, 506, 372, 483, 576, 48]);
+    });
+
+    it('encoded numbers should by in [0, 1024)', () => {
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/services/SecurityCodeService.ts]: ( -> ]

**Pull request merged**